### ISSUE TITLE
reg-tests/compression/lua_validation.vtc   fails on Fedora Rawhide

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#1141]
```
########################## Starting vtest ##########################
Testing with haproxy version: 2.4-dev8-5567f4-38
#    top  TEST reg-tests/compression/lua_validation.vtc FAILED (0.512) exit=2
1 tests failed, 0 tests skipped, 122 tests passed
########################## Gathering results ##########################
###### Test case: reg-tests/compression/lua_validation.vtc ######
## test results in: "/tmp/haregtests-2021-02-19_16-04-56.f2aNi0/vtc.5140.2db66cc5"
---- top   shell_exit not as expected: got 0x0001 wanted 0x0000
make: *** [Makefile:1059: reg-tests] Error 1
Test case: reg-tests/compression/lua_validation.vtc
**** dT    0.000
*    top   TEST reg-tests/compression/lua_validation.vtc starting
**** top   extmacro def pwd=/builds/chipitsine/haproxy-ci-playground
**** top   extmacro def no-htx=
**** top   extmacro def localhost=127.0.0.1
**** top   extmacro def bad_backend=127.0.0.1:43789
**** top   extmacro def listen_addr=127.0.0.1:0
**** top   extmacro def bad_ip=192.0.2.255
**** top   macro def testdir=/builds/chipitsine/haproxy-ci-playground/reg-tests/compression
**** top   macro def tmpdir=/tmp/haregtests-2021-02-19_16-04-56.f2aNi0/vtc.5140.2db66cc5
**   top   === varnishtest "Compression validation"
*    top   VTEST Compression validation
**   top   === feature ignore_unknown_macro
**   top   === haproxy h1 -conf {
**** h1    macro def h1_closed_sock=127.0.0.1 42735
**** h1    macro def h1_closed_addr=127.0.0.1
**** h1    macro def h1_closed_port=42735
**** dT    0.004
**** h1    macro def h1_cli_sock=127.0.0.1 38285
**** h1    macro def h1_cli_addr=127.0.0.1
**** h1    macro def h1_cli_port=38285
**** h1    setenv(cli, 4)
**** h1    macro def h1_fe1_sock=127.0.0.1 35865
**** h1    macro def h1_fe1_addr=127.0.0.1
**** h1    macro def h1_fe1_port=35865
**** h1    setenv(fe1, 5)
**** h1    macro def h1_fe2_sock=127.0.0.1 40785
**** h1    macro def h1_fe2_addr=127.0.0.1
**** h1    macro def h1_fe2_port=40785
**** h1    setenv(fe2, 6)
**   h1    haproxy_start
**** h1    opt_worker 0 opt_daemon 0 opt_check_mode 0 opt_mcli 0
**** h1    argv|exec "/builds/chipitsine/haproxy-ci-playground/haproxy" -d -dM  -f "/tmp/haregtests-2021-02-19_16-04-56.f2aNi0/vtc.5140.2db66cc5/h1/cfg" 
**** h1    conf|    global
**** h1    conf|\tstats socket "/tmp/haregtests-2021-02-19_16-04-56.f2aNi0/vtc.5140.2db66cc5/h1/stats.sock" level admin mode 600
**** h1    conf|    stats socket "fd@${cli}" level admin
**** h1    conf|
**** h1    conf|global
**** h1    conf|#\tlog stdout format short daemon
**** h1    conf|\tlua-load\t\t/builds/chipitsine/haproxy-ci-playground/reg-tests/compression/lua_validation.lua
**** h1    conf|
**** h1    conf|defaults
**** h1    conf|\tmode\t\t\thttp
**** h1    conf|\tlog\t\t\tglobal
**** h1    conf|\t option http-use-htx
**** h1    conf|\toption\t\t\thttplog
**** h1    conf|
**** h1    conf|frontend main-https
**** h1    conf|\tbind\t\t\t"fd@${fe1}" ssl crt /builds/chipitsine/haproxy-ci-playground/reg-tests/compression/common.pem
**** h1    conf|\tcompression algo gzip
**** h1    conf|\tcompression type text/html text/plain application/json application/javascript
**** h1    conf|\tcompression offload
**** h1    conf|\tuse_backend TestBack  if  TRUE
**** h1    conf|
**** h1    conf|backend TestBack
**** h1    conf|\tserver\tLocalSrv 127.0.0.1:40785
**** h1    conf|
**** h1    conf|listen fileloader
**** h1    conf|\tmode http
**** h1    conf|\tbind "fd@${fe2}"
**** h1    conf|\thttp-request use-service lua.fileloader-http01
**** h1    XXX 8 @735
***  h1    PID: 5745
**** h1    macro def h1_pid=5745
**** h1    macro def h1_name=/tmp/haregtests-2021-02-19_16-04-56.f2aNi0/vtc.5140.2db66cc5/h1
**   top   === shell {
**** top   shell_cmd|exec 2>&1 ; 
**** top   shell_cmd|    HOST=127.0.0.1
**** top   shell_cmd|    if [ "127.0.0.1" = "::1" ] ; then
**** top   shell_cmd|        HOST="\[::1\]"
**** top   shell_cmd|    fi
**** top   shell_cmd|
**** top   shell_cmd|    md5=$(command -v md5 || command -v md5sum)
**** top   shell_cmd|
**** top   shell_cmd|    if [ -z $md5 ] ; then
**** top   shell_cmd|        echo "MD5 checksum utility not found"
**** top   shell_cmd|        exit 1
**** top   shell_cmd|    fi
**** top   shell_cmd|
**** top   shell_cmd|    expectchecksum="4d9c62aa5370b8d5f84f17ec2e78f483"
**** top   shell_cmd|
**** top   shell_cmd|    for opt in "" "--limit-rate 300K" "--limit-rate 500K" ; do
**** top   shell_cmd|        checksum=$(curl --max-time 15 --compressed -k "https://$HOST:35865" $opt | $md5 | cut -d ' ' -f1)
**** top   shell_cmd|        if [ "$checksum" != "$expectchecksum" ] ; then
**** top   shell_cmd|              echo "Expecting checksum $expectchecksum"
**** top   shell_cmd|              echo "Received checksum: $checksum"
**** top   shell_cmd|              exit 1;
**** top   shell_cmd|        fi
**** top   shell_cmd|    done
**** top   shell_cmd|
**** dT    0.282
***  h1    debug|[WARNING] 049/160516 (5745) : config : missing timeouts for frontend 'main-https'.
***  h1    debug|   | While not properly invalid, you will certainly encounter various problems
***  h1    debug|   | with such a configuration. To fix this, please ensure that all following
***  h1    debug|   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
***  h1    debug|[WARNING] 049/160516 (5745) : config : log format ignored for frontend 'main-https' since it has no log address.
***  h1    debug|[WARNING] 049/160516 (5745) : config : missing timeouts for backend 'TestBack'.
***  h1    debug|   | While not properly invalid, you will certainly encounter various problems
***  h1    debug|   | with such a configuration. To fix this, please ensure that all following
***  h1    debug|   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
**** dT    0.283
***  h1    debug|[WARNING] 049/160516 (5745) : config : missing timeouts for proxy 'fileloader'.
***  h1    debug|   | While not properly invalid, you will certainly encounter various problems
***  h1    debug|   | with such a configuration. To fix this, please ensure that all following
***  h1    debug|   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
***  h1    debug|[WARNING] 049/160516 (5745) : config : log format ignored for proxy 'fileloader' since it has no log address.
**** dT    0.285
***  h1    debug|Note: setting global.maxconn to 524274.
***  h1    debug|Available polling systems :
***  h1    debug|      epoll : 
***  h1    debug|pref=300, 
***  h1    debug| test result OK
***  h1    debug|
***  h1    debug|       poll : 
***  h1    debug|pref=200, 
***  h1    debug| test result OK
***  h1    debug|
***  h1    debug|     select : 
***  h1    debug|pref=150, 
***  h1    debug| test result FAILED
***  h1    debug|
***  h1    debug|Total: 3 (2 usable), will use epoll.
***  h1    debug|
***  h1    debug|Available filters :
***  h1    debug|\t[SPOE] spoe
***  h1    debug|\t[CACHE] cache
***  h1    debug|\t[FCGI] fcgi-app
***  h1    debug|\t[COMP] compression
***  h1    debug|\t[TRACE] trace
**** dT    0.301
***  h1    debug|Using epoll() as the polling mechanism.
**** dT    0.329
***  h1    debug|00000000:main-https.accept(0005)=000b from [127.0.0.1:33416] ALPN=<none>
***  h1    debug|00000000:main-https.clireq[000b:ffffffff]: GET / HTTP/1.1
***  h1    debug|00000000:main-https.clihdr[000b:ffffffff]: host: 127.0.0.1:35865
***  h1    debug|00000000:main-https.clihdr[000b:ffffffff]: user-agent: curl/7.75.0
***  h1    debug|00000000:main-https.clihdr[000b:ffffffff]: accept: */*
***  h1    debug|00000000:main-https.clihdr[000b:ffffffff]: accept-encoding: deflate, gzip, br
***  h1    debug|00000001:fileloader.accept(0006)=000d from [127.0.0.1:40096] ALPN=<none>
***  h1    debug|00000001:fileloader.clireq[000d:ffffffff]: GET / HTTP/1.1
***  h1    debug|00000001:fileloader.clihdr[000d:ffffffff]: host: 127.0.0.1:35865
***  h1    debug|00000001:fileloader.clihdr[000d:ffffffff]: user-agent: curl/7.75.0
***  h1    debug|00000001:fileloader.clihdr[000d:ffffffff]: accept: */*
***  h1    debug|00000001:fileloader.srvrep[000d:ffffffff]: HTTP/1.1 200 OK
***  h1    debug|00000001:fileloader.srvhdr[000d:ffffffff]: content-type: application/javascript
***  h1    debug|00000001:fileloader.srvhdr[000d:ffffffff]: content-length: 1888190
**** dT    0.330
***  h1    debug|00000000:TestBack.srvrep[000b:000c]: HTTP/1.1 200 OK
***  h1    debug|00000000:TestBack.srvhdr[000b:000c]: content-type: application/javascript
***  h1    debug|00000000:TestBack.srvhdr[000b:000c]: content-length: 1888190
**** dT    0.344
***  h1    debug|00000001:fileloader.srvcls[000d:ffffffff]
***  h1    debug|00000001:fileloader.clicls[000d:ffffffff]
***  h1    debug|00000001:fileloader.closed[000d:ffffffff]
**** dT    0.377
***  h1    debug|00000000:TestBack.srvcls[000b:000c]
***  h1    debug|00000000:TestBack.clicls[000b:000c]
***  h1    debug|00000000:TestBack.closed[000b:000c]
**** dT    0.397
***  h1    debug|00000002:main-https.accept(0005)=000b from [127.0.0.1:33420] ALPN=<none>
***  h1    debug|00000002:main-https.clireq[000b:ffffffff]: GET / HTTP/1.1
***  h1    debug|00000002:main-https.clihdr[000b:ffffffff]: host: 127.0.0.1:35865
***  h1    debug|00000002:main-https.clihdr[000b:ffffffff]: user-agent: curl/7.75.0
***  h1    debug|00000002:main-https.clihdr[000b:ffffffff]: accept: */*
***  h1    debug|00000002:main-https.clihdr[000b:ffffffff]: accept-encoding: deflate, gzip, br
***  h1    debug|00000003:fileloader.accept(0006)=000f from [127.0.0.1:40100] ALPN=<none>
***  h1    debug|00000003:fileloader.clireq[000f:ffffffff]: GET / HTTP/1.1
***  h1    debug|00000003:fileloader.clihdr[000f:ffffffff]: host: 127.0.0.1:35865
***  h1    debug|00000003:fileloader.clihdr[000f:ffffffff]: user-agent: curl/7.75.0
***  h1    debug|00000003:fileloader.clihdr[000f:ffffffff]: accept: */*
***  h1    debug|00000003:fileloader.srvrep[000f:ffffffff]: HTTP/1.1 200 OK
***  h1    debug|00000003:fileloader.srvhdr[000f:ffffffff]: content-type: application/javascript
***  h1    debug|00000003:fileloader.srvhdr[000f:ffffffff]: content-length: 1888190
***  h1    debug|00000002:TestBack.srvrep[000b:000e]: HTTP/1.1 200 OK
***  h1    debug|00000002:TestBack.srvhdr[000b:000e]: content-type: application/javascript
***  h1    debug|00000002:TestBack.srvhdr[000b:000e]: content-length: 1888190
**** dT    0.406
***  h1    debug|00000003:fileloader.srvcls[000f:ffffffff]
***  h1    debug|00000003:fileloader.clicls[000f:ffffffff]
***  h1    debug|00000003:fileloader.closed[000f:ffffffff]
**** dT    0.411
**** top   shell_out|  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
**** top   shell_out|                                 Dload  Upload   Total   Spent    Left  Speed
**** top   shell_out|\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r100  418k    0  418k    0     0  1208k      0 --:--:-- --:--:-- --:--:-- 1204k
**** top   shell_out|  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
**** top   shell_out|                                 Dload  Upload   Total   Spent    Left  Speed
**** top   shell_out|\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r100 10418    0 10418    0     0   462k      0 --:--:-- --:--:-- --:--:--  462k
**** top   shell_out|curl: (61) Error while processing content unencoding: invalid code lengths set
**** top   shell_out|Expecting checksum 4d9c62aa5370b8d5f84f17ec2e78f483
**** top   shell_out|Received checksum: 4e88a22c8fa5729ceab7fd0957f294cb
**** top   shell_status = 0x0001
---- top   shell_exit not as expected: got 0x0001 wanted 0x0000
*    top   RESETTING after reg-tests/compression/lua_validation.vtc
**   h1    Reset and free h1 haproxy 5745
**   h1    Wait
**   h1    Stop HAproxy pid=5745
**** h1    Kill(2)=0: Success
**** dT    0.412
**** h1    STDOUT EOF
**** dT    0.511
**   h1    WAIT4 pid=5745 status=0x0002 (user 0.195154 sys 0.142395)
*    top   TEST reg-tests/compression/lua_validation.vtc FAILED
```

**Comment by chipitsine** [comment]
failed build:
https://gitlab.com/chipitsine/haproxy-ci-playground/-/jobs/1042637796

**Comment by a-denoyelle** [comment]
so it seems I'm not the only one to reproduce it now (archlinux for me) :)

**Comment by capflam** [comment]
I remember @a-denoyelle already show me such failures. But I'm unable to reproduce it. And it seems a bit flaky.

**Comment by chipitsine** [comment]
![Screenshot from 2021-02-19 22-21-46](https://user-images.githubusercontent.com/2217296/108538763-f7d41e00-7300-11eb-9324-6e424ff98175.png)


it changed today from "all green" to "all red"

**Comment by wtarreau** [comment]
It's all green again.

**Comment by capflam** [comment]
It starts to fail on a change in the DOC ...  @wtarreau, not on gitlab. The result is not reported on the GH interface (AFAIK)

**Comment by a-denoyelle** [comment]
Personnally I can reproduce it every time, I'm investigating it.

**Comment by capflam** [comment]
@chipitsine but I would be happy to understand why it fails. Thanks @a-denoyelle !

**Comment by wtarreau** [comment]
@capflam on what gitlab ? I think I'm totally lost with all these tools around :-/

**Comment by TimWolla** [comment]
> @capflam on what gitlab ? I think I'm totally lost with all these tools around :-/

Ilya appears to run a private fork on GitLab.com: https://gitlab.com/chipitsine/haproxy-ci-playground/-/pipelines.

This private fork is fully independent from the official GitHub repository.

**Comment by chipitsine** [comment]
I think "something changed in Fedora Rawhide".
it might be whatever: curl, lua, whatever

**Comment by chipitsine** [comment]
we can add Fedora Rawhide builds to GH Actions if we want to

**Comment by a-denoyelle** [comment]
for me the test works with curl 7.74 but starts to fail with 7.75

**Comment by chipitsine** [comment]
it is not a private fork. it is automated fork

![Screenshot from 2021-02-19 23-16-43](https://user-images.githubusercontent.com/2217296/108544721-b21b5380-7308-11eb-9318-5121a8271fa7.png)

**Comment by a-denoyelle** [comment]
I bisected curl and for me the guilty commit is this one : 
```
commit b68dc34af341805aeb7b371541a2b4074da76217 (refs/bisect/bad)
Date:   Thu Jan 14 14:24:07 2021 +0100
multi: set the PRETRANSFER time-stamp when we switch to PERFORM
```

This looks like a bug in curl because it seems to be an internal change. I will report the issue when I can find an easily reproducible scenario.

**Comment by TimWolla** [comment]
> we can add Fedora Rawhide builds to GH Actions if we want to

@chipitsine In fact I looked into that in the last days. However it was blocked on https://github.com/fedora-cloud/docker-brew-fedora/issues/86. I'll retest that in the next days and prepare a patch.

**Comment by chipitsine** [comment]
I use the following script:
https://gitlab.com/chipitsine/haproxy-playground-scripts/-/blob/master/.external.yml

(not sure why do they ship image without gpg keys)

**Comment by a-denoyelle** [comment]
Also note that the issue is not reproducible with wget, this might be an alternative for the test.

```
checksum=$(wget -O - --compression=gzip --no-check-certificate $opt "https://$HOST:${h1_fe1_port}" | $md5 | cut -d ' ' -f1)
```

I have opened an issue on the curl github repository.

**Comment by TimWolla** [comment]
> I have opened an issue on the curl github repository.

For reference: curl/curl#6640

**Comment by a-denoyelle** [comment]
A fix was pushed on curl which fixes the issue.

```
commit bf601472a0f09ff5fa80a048f22a30ac9edac304 (HEAD, origin/master, origin/HEAD)
multi: do once-per-transfer inits in before_perform in DID state
```

I confirm this fixes the regtest, so only curl 7.75 will cause a failure. We can close this issue for me.
